### ISSUE TITLE
Video Data augmentation & Preprocessing Pipelines - Augmentations

### ISSUE THREAD

**Issue by Tony363** [keras-team/keras-cv#311]
**A. Robust, portable, and simple Video Augmentation wrapper** 
1. Collect, document, and find known Video augmentation libraries
2. Design Keras Video Augmentation Wrapper API
3. Construct bare-bones objects and methods with ease of use in mind
4. Refactor code for flexible user input “augmentors” to wrapper
5. Testing, Documenting, confirming the robustness of wrapper
[Proposal link:](https://docs.google.com/document/d/1_xcWXFQjekvtwrhsqaZN03ribcD65-GWyNZ-EgSfJrA/edit?usp=sharing) 
Please comment on proposal

**Comment by sayakpaul** [comment]
Definitely following this one. I think before writing the necessary augmentation packages we need a good data pipeline that allows users to efficiently load videos and feed them to models. 

I wrote two examples covering videos but I am not super happy about the data pipeline:

* https://keras.io/examples/vision/video_classification/
* https://keras.io/examples/vision/video_transformers/

**Comment by Tony363** [comment]
My thoughts exactly as well! However, it may be easier for me to start my Google summer of code writer the video augmented wrapper. But yes, a well designed project will start engineering from the pipeline itself! I was about to fill in the details of my GSOC proposal on the Video data pipeline today or tomorrow !

**Comment by bhack** [comment]
Without the opencv dependency we could use also https://github.com/tensorflow/io/blob/master/tensorflow_io/python/experimental/video_dataset_ops.py

**Comment by Tony363** [comment]
> Without the opencv dependency we could use also https://github.com/tensorflow/io/blob/master/tensorflow_io/python/experimental/video_dataset_ops.py

I'll read up on this and add it to my proposal. However, from my work experience, most vision APIs draw from opencv underneath 1 way or another which is drawn from its cpp source code. Just curious, are there specific detriments to using opencv? I myself faced alot opencv contribs, extras with cuda/cudnn cross platform compilation issues in the past but vanilla opencv from pip should be fine? Maybe the design architecture if python being a cpp API doesnt fit into keras?

**Comment by Tony363** [comment]
> * https://keras.io/examples/vision/video_classification/

Love those examples you wrote by the way! I currently in the process of optimizing my own video data pipeline for my research in collaboration a Google Pixel project! Your examples have direct applications for me and Ill definitely peruse through them and add them to my Google summer of code proposal!

**Comment by bhack** [comment]
I suppose that we don't want to overlap/duplicate too much on what it is already working the TF-IO SIG /cc @yongtang and to the Video dataset adapters that we have at https://www.tensorflow.org/datasets/api_docs/python/tfds/features/Video

**Comment by Tony363** [comment]
> Sure. Thank you.
> 
> Let's maybe collaborate on creating an efficient video data pipeline which the users could refer to.
> 
> The goal would be to write a data pipeline (preferably tf.data) and show how to feed that to a video model.

This is a [pipeline](https://github.com/hc07180011/testing-cv/blob/main/flicker_detection/flicker_detection/train.py) that I wrote that does train test split based on file names then populates the padded batched chunks into a tuple of chunks. These are all done in a single loop. Before it was using lists, which required unnecessary RAM memory and dimensions to each single Batched chunk. Through out the preprocessing, I would have write to file check points so as to not have to restart from an already finished process. However, as it stands, this approach is far too unstandardized for me to analyze the batches itself and how they correspond to my labels. In this example, I managed to reduce train and test data cache from 5.93 GB to 2.63 GB. Its difficult for me to figure out the data integrity with this kind of optimization tho. 

I thought about using tf.Data, as well as Ragged tensors, and tf.Datas TFRecords with data bytecode serialization. I think an important feature would be how to load the data into GPU memory and performing error analysis with the tensors. Another feature would be to effectively utilize static memory and not entirely rely an RAM memory.

**Comment by sayakpaul** [comment]
>  I think an important feature would be how to load the data into GPU memory and performing error analysis with the tensors. Another feature would be to effectively utilize static memory and not entirely rely an RAM memory.

Hmm. I think one also needs to be careful about using GPU memory for data processing when training models in conjunction. There can be a lot of things happening there -- model states, optimizer states, temporary buffers, unusable memory due fragmentation, etc. So, naturally, processing video data with GPU memory could be more challening than it might sound. 

But I am curious. 

@bhack would be nice to have clear and concise tutorials / guides that commodity users can refer to for videos. Not talking about reading source code or documentation here.

**Comment by Tony363** [comment]
> > I think an important feature would be how to load the data into GPU memory and performing error analysis with the tensors. Another feature would be to effectively utilize static memory and not entirely rely an RAM memory.
> 
> Hmm. I think one also needs to be careful about using GPU memory for data processing when training models in conjunction. There can be a lot of things happening there -- model states, optimizer states, temporary buffers, unusable memory due fragmentation, etc. So, naturally, processing video data with GPU memory could be more challening than it might sound.
> 
> But I am curious.
> 
> @bhack would be nice to have clear and concise tutorials / guides that commodity users can refer to for videos. Not talking about reading source code or documentation here.

I think It may also be beneficial to appropriately cache resources into static memory at key pipeline intervals and to properly seed these preprocessing checkpoints. Maybe when working with video data sets, it is preferential to preprocess and train models separately?

**Comment by bhack** [comment]
What I meant:

Tutorials may or may not use ecosystem synergies. I think the important thing is not to reinvent the wheel in the ecosystem, users are not very interested in duplications, on the contrary they are often confused when there are several ways to do the same thing within the ecosystem.

**Comment by sayakpaul** [comment]
> Tutorials may or may not use ecosystem synergies. I think the important thing is not to reinvent the wheel in the ecosystem, users are not very interested in duplications, on the contrary they are often confused when there are several ways to do the same thing within the ecosystem.

I am asking for official tutorials using the components you mentioned.

**Comment by sayakpaul** [comment]
> Maybe when working with video data sets, it is preferential to preprocess and train models separately?

Presumably. That is why I tried following a combined approach in my guides.

**Comment by Tony363** [comment]
@sayakpaul Whatever the case, Im putting all this into my proposal XD

**Comment by Tony363** [comment]
> 

On the subject of fragmentation, @sayakpaul do you think its feasible to integrate a garbage collector specifically for Video data preprocessing. One thing I always do to make sure my preprocessing runs and is able to feed to my training function is to constantly collect unused memory or delete objects that are either no longer used or have been written to file. Otherwise, a simple mishap and my google colab note book immediately crashes.

... and ya, im currently only making by with google collaborator and don't have a machine of my own capable of handling such heavy payloads

**Comment by sayakpaul** [comment]
My analysis of the consumption of GPU memory stems from ZeRO from Microsoft. I am not too sure if garbage collection would cut it but worth exploring.

**Comment by Tony363** [comment]
> My analysis of the consumption of GPU memory stems from ZeRO from Microsoft. I am not too sure if garbage collection would cut it but worth exploring.

After briefly reading the abstract of ZeRO, I think I understand your perspective. I will peruse through the document of ZeRO as well and think about how best to tackle preprocessing with the most resourceful approach!

**Comment by Tony363** [comment]
@sayakpaul Hello again. I finished drafting the pipeline portion of this issue, "Video Data augmentation & Preprocessing Pipelines - Augmentations", for Google Summer of Code. I am now moving on to draft the portion of this issue on the theme of integrating the augmentor wrapper and the pipeline wrapper together with the main focus of tackling RAM memory requirements to make the final module a light weight, easy to use, customizable module. You previously [mentioned](https://github.com/keras-team/keras-cv/issues/311#issuecomment-1099221225) in a comment above that utilizing GPU memory for preprocessing would be more challenging that it may be.

This being so, how might a general outline to tackle this problem look like?

**Comment by sayakpaul** [comment]
Thanks for further thinking through this. 

Honestly, I am myself a bit hazy about how does an efficient pipeline look like in this case. I will have consult libraries like `scenic` and also what @bhack shared to develop a general idea.

**Comment by Tony363** [comment]
@sayakpaul Shall we leave the final section of the proposal as an "agile" development approach where we leave the remaining time of the GSOC to optimize when issues come up? Because maybe some issues can only be understand hence getting our hands dirty?

**Comment by sayakpaul** [comment]
Sounds good to me.

**Comment by Tony363** [comment]
@sayakpaul Sorry for asking again. I contemplated whether to add an embedding visualization aspect to the Video preprocessing data pipeline. Would you know of any existent keras native embedding visualization tools? Im exploring tensorboard right now, but it seems most of its feature are for training model callbacks?

Im also wondering if there are ways to speed up feature extraction from base cnns like mobilenet.MobileNet. An issue that i am facing right at this moment is that i have 1991 number of videos to extract embeddings from.
 ` self.__embedding(resnet.preprocess_input(image_tensor)).numpy()`, this piece is the main embedding extraction code piece.. However, Ive already spend 7-8 hours extracting embeddings, and ive only extracted around 160 of 1991.

**Comment by Tony363** [comment]
I think i finished my rough draft of my proposal. Please check it out~

**Comment by Tony363** [comment]
@sayakpaul Just submited my GSOC proposal! See you on the other end of things or I might as well start contributing now since Ill be working on my research project data pipeline anyways~

**Comment by Tony363** [comment]
@sayakpaul  Hey, its been a while. I had Midterms the past week.  I'm rewriting a new pipeline for video classification problems and borrowed a lot of inspiration from your tutorials. Here is the [feature extraction script](https://github.com/hc07180011/testing-cv/blob/main/flicker_detection/flicker_detection/extract_embedding.py) I have so far.

These are the [BaseCNN & Serializer](https://github.com/hc07180011/testing-cv/blob/main/flicker_detection/flicker_detection/preprocessing/embedding/backbone.py) object classes that facilitate the scripts. 

I'm currently facing a couple of issues after serializing my embeddings into TFRecords. Basically, I serialize video embeddings from batches of 32 frames as a single .tfrecords while recording its keys into a .json file. I am trying to decode the .tfrecords but I'm having an issue. 

**[tensorflow.python.framework.er](http://tensorflow.python.framework.er/)rors_[impl.In](http://impl.in/)validArgumentError: Key: batch_56. Data types don't match. Data type: string but expected type: float [Op:ParseExampleV2]** <- this is the error I'm having.

The general design of the feature extraction/preprocessing script is to extract embeddings by batches while also padding for assymetric batches. Do a train test split on the .tfrecord files in its saved directory and then repopulate them as required in the training loop. I'm stuck with figuring out how to decode the .tfrecords to repopulate the training tensors on demand. 

Other than the above issue. I tried a workaround via saving the batched embeddings as .npy, however, that would have the operating system hang between GPU embedding inference and CPU numpy save to file, a memory thrashing issue?

I tried out your code to implement a CNN-Transformers model for video classification. My model went from 19,000,000 parameters to 340,000,000 parameters with a ResourceExhaust error..... I think I really need to optimize my data pipeline before I can further try additional experiments.

**Comment by LukeWood** [comment]
The KerasCV team is attempting to narrow the package scope in order to more efficiently use our limited resources. As such, we are only accepting contributions directly related to items on the 6 month roadmap for the package.

If you feel this feature could be particularly useful, or would like to contribute it - please feel free to re-open this issue and ping LukeWood, ianstenbit or tanzhenyu

Feel free to read more on the roadmap:
https://github.com/keras-team/keras-cv/blob/master/.github/ROADMAP.md
